feat(#161): showdown signature verification

augustin-v opened the pull request:
## Description   
Implemented signature verification in the showdown function to ensure that each player's hand is authorized by their ECDSA signature. This includes recovering the public key from the signature, comparing it with the stored public key, verifying the player is part of the current game round, and preventing replay attacks with a nonce.

Testing turned out to be complex due to the need for valid signatures on the secp256r1 curve. A comprehensive test suite with proper signature generation is planned for a follow-up PR.



## Related Issues   
Closes #161

## Changes Made
- Added signature parameters (r, s arrays, y_parity, nonce) to showdown function
- Computed Poseidon hash of hand + nonce as message hash for signature verification
- Implemented public key recovery from signature using [`recover_public_key`](https://docs.cairo-lang.org/core/core-ecdsa-recover_public_key.html) (same name but different from the one in the issue #161 where it returns a point on the curve[this one](https://docs.cairo-lang.org/core/core-starknet-secp256_trait-recover_public_key.html) )
- Added comparison between recovered public key and stored player public key
- Verified player participation in current game round
- Incremented nonce to prevent replay attacks and ensure game state persistence
- Added additional ECDSA signature verification using `check_ecdsa_signature`
- Implemented comprehensive error handling and input validation
- Added array bounds safety checks for signature parameters
- Updated Player model to include pub_key field for signature verification

## How to Test  
- Deploy the contract and call the showdown function with valid hands and corresponding signatures
- Ensure that invalid signatures or mismatched nonces cause the function to revert with appropriate error messages
- Verify that the nonce increments after a successful showdown call
- Test with mismatched signature array lengths to verify input validation
- Confirm that players not in the current round are rejected


## Screenshots (if applicable)  
N/A

## Checklist  
- [x] My code follows the project's coding style
- [x] I have tested these changes locally (manual verification)
- [ ] Documentation has been updated where necessary

augustin-v committed:
feat(#161): showdown signature verification

Birdmannn commented on poker-texas-hold-em/contract/src/systems/actions.cairo:
```diff
@@ -279,19 +285,80 @@ pub mod actions {
+                assert(pubkey_felt == player.pub_key, 'INVALID PUB KEY');
+
+                // verify player is in current game
+                let mut player_found: bool = false;
+                let mut j: u32 = 0;
+                while j < game.players.len() {
+                    if *game.players.at(j) == *hand.player {
+                        player_found = true;
+                        break;
+                    }
+                    j += 1;
+                };
```
Please just assert `player.in_round` and `player.is_in_game(game_id)` is true. 
so this block should be removed, and the line following this block should be

`assert(player.in_round && player.is_in_game(game_id), 'PLAYER NOT IN ROUND')`

Why?
It reduces the overhead loop, and the game.players value is optimized using a certain logic and criteria, so it is not always correct.

Birdmannn replied:
And please, resolve conflicts 🙂 @augustin-v

Birdmannn reviewed (commented):
Sorry, in your description, you said you used this:

```
use starknet::secp256r1::Secp256r1Point;
use starknet::secp256_trait::{Signature, recover_public_key};
```

but in the code, you didn't later use it. Any reason for the change of mind, sir?

augustin-v commented:
Thank you for the review @Birdmannn.



> Sorry, in your description, you said you used this:
> 
> ```
> use starknet::secp256r1::Secp256r1Point;
> use starknet::secp256_trait::{Signature, recover_public_key};
> ```
> 
> but in the code, you didn't later use it. Any reason for the change of mind, sir?

Sorry for the unclear description, I also forgot to remove these imports

https://github.com/Prometheus-A/The-Velvet-Ace/blob/333c5dc3eca21175abe2175f7142f3de0aa236d9/poker-texas-hold-em/contract/src/systems/actions.cairo#L9-L10

Here is what happened during development:

### Initial Plan:
- Use `starknet::secp256_trait::recover_public_key` which returns `Option<Secp256r1Point>`
- Store public keys as `Secp256r1Point` in the Player model
- Compare recovered points directly with stored points

Why I Changed Approach:
1. Discovered `core::ecdsa::recover_public_key`: This returns `Option<felt252>` (x-coordinate, later used for the check_ecdsa_signature) instead of a full point, which is simpler to work with
2. Type compatibility: `check_ecdsa_signature` expects `felt252` public keys, not points
3. Simplified storage: Storing public keys as `felt252` is more efficient than storing curve points

Final Implementation:
- Used `core::ecdsa::recover_public_key` (returns felt252)
- Stored public keys as `felt252` in Player model
- Direct comparison between recovered and stored public keys
- Compatible with both recovery and `check_ecdsa_signature` verification

What do you think, does this make sense or should we go another approach?

Birdmannn commented:
> What do you think, does this make sense or should we go another approach?

It should make sense... unless you have another approach 😅
This is the only approach I know of for now.

Birdmannn reviewed (approved):
Trés bien! @augustin-v. As good as I anticipated... After all it's you sir 🫡.
I really hope we meet again, whatever project it may be.

Thank you very much 🙏🏻

augustin-v commented:
Merci @Birdmannn ! Of course looking forward to work with you again
